### DISCUSSION TITLE
[Question] Is there any way to save a  trained model in pytorch tabular

### DISCUSSION THREAD

**Discussion by pranavvp16** [manujosephv/pytorch_tabular#272] [General]
Currently I'm working on task that requires to save the model, can I save the model using pickle or other libraries like dill

**Comment by DonMiller9294** [comment]
Absolutely, you can save your Python model using libraries like `pickle` or even `dill`. They're handy tools for serializing Python objects, including machine learning models, and storing them for future use.

Here's a quick example of how you can save a model using `pickle`:

```python
import pickle

# Assuming you have a trained model
model = ...

# Now, let's save this model to a file
with open('model.pkl', 'wb') as file:
    pickle.dump(model, file)
```

And when you need to bring that model back to life, just use `pickle` again:

```python
import pickle

# Loading the model from the saved file
with open('model.pkl', 'rb') as file:
    loaded_model = pickle.load(file)
```

It's a simple and nifty way to save and reload your models. However, do keep in mind that for complex machine learning models, especially those from deep learning frameworks like PyTorch or TensorFlow, using the built-in serialization methods provided by the framework (like `torch.save` and `torch.load` for PyTorch) might be a more reliable choice, as they handle the model's architecture and parameters more effectively. Feel free to ask if you have more questions or need further assistance! 😊

**Comment by pranavvp16** [comment]
@DonMiller9294 thanks for your reply, last question how can I save a pytorch tabular model using `torch.save` as mentioned above. If this possible can I make PR in the examples section which also saves the model. I think I would be handy for future contributors

**Comment by manujosephv** [comment]
@pranavvp16 Everything that @DonMiller9294 said works. For added convenience, PyTorch Tabular provides a few methods out of the box to help you save and load the model.

1. [save_model](https://pytorch-tabular.readthedocs.io/en/latest/apidocs_coreclasses/#pytorch_tabular.tabular_model.TabularModel.save_model) - This does multiple things: pickles the config, datamodule, saves the model checkpoint, also any custom parameters you have passed in while running the model in a standard way so that at a later point of time you can load the model fully using [load_model](https://pytorch-tabular.readthedocs.io/en/latest/apidocs_coreclasses/#pytorch_tabular.tabular_model.TabularModel.load_model). This makes the TabularModel apis like predict etc. work while inference.
3. If you want to do the data processing for inference at your side, PyTorch Tabular also provides a way to just save the state dict of the model and load them using these methods: [save_weights](https://pytorch-tabular.readthedocs.io/en/latest/apidocs_coreclasses/#pytorch_tabular.tabular_model.TabularModel.save_weights) and [load_weights](https://pytorch-tabular.readthedocs.io/en/latest/apidocs_coreclasses/#pytorch_tabular.tabular_model.TabularModel.load_weights). This is more useful if you want to reset the weights while runtime or move to another loaded state dict etc. Under this method, responsibility of initializing the right model with the same parameters etc is upto the user,
4. There is another method [save_model_for_inference](https://pytorch-tabular.readthedocs.io/en/latest/apidocs_coreclasses/#pytorch_tabular.tabular_model.TabularModel.save_model_for_inference) which saves the whole model using `torch.save`. It also has experimental support to export to `ONNX`. This acts as a pure pytorch model and while inference user would have to ensure that the data is processed the right way, same transformations etc and fed to the model as a tensor.


And it would be absolutely wonderful if you can add examples or add documentation to make this process clearer..

**Comment by pranavvp16** [comment]
Thanks, I will make a PR soon

**Comment by manujosephv** [comment]
Converting this to a discussion, as this isn't an issue..
